tests/strategies/test_ddp_strategy.py

```diff
@@ -127,3 +133,48 @@ def test_ddp_configure_ddp():
     trainer.strategy.setup(trainer)
     # in DDPStrategy configure_ddp(), model are still LightningModule
     assert isinstance(trainer.model, LightningModule)
+
+
+class BoringFairScaleOptimizerModel(BoringModel):
+    def configure_optimizers(self):
+        base_optimizer = torch.optim.SGD(self.layer.parameters(), lr=0.1)
+        return OSS(params=base_optimizer.param_groups, optim=type(base_optimizer), **base_optimizer.defaults)
+
+
+@RunIf(min_gpus=2, skip_windows=True, fairscale=True)
+def test_ddp_strategy_checkpoint_multi_gpu_fairscale_optimizer(tmpdir):
+    """Test to ensure that checkpoint is saved correctly when using faircale optimizer."""
+    model = BoringFairScaleOptimizerModel()
+    trainer = Trainer(gpus=2, max_epochs=2, strategy="ddp")
+
+    trainer.fit(model)
+
+    checkpoint_path = os.path.join(tmpdir, "model.pt")
+    trainer.save_checkpoint(checkpoint_path)
+    saved_model = BoringModel.load_from_checkpoint(checkpoint_path)
+
+    # Assert model parameters are identical after loading
+    for ddp_param, shard_param in zip(model.parameters(), saved_model.parameters()):
+        assert torch.equal(ddp_param.to("cpu"), shard_param)
+
+
+class BoringZeroRedundancyOptimizerModel(BoringModel):
+    def configure_optimizers(self):
+        return ZeroRedundancyOptimizer(self.layer.parameters(), optimizer_class=torch.optim.Adam, lr=0.1)
+
+
+@RunIf(min_torch="1.10")
+def test_ddp_strategy_checkpoint_zero_redundancy_optimizer(tmpdir):
```

Comment:
this test seems not working on multi-gpu with errors 
```
pytorch_lightning/strategies/strategy.py:166: in optimizer_state
    optimizer.consolidate_state_dict()
../local/miniconda3/envs/pl-fariscale/lib/python3.9/site-packages/torch/distributed/optim/zero_redundancy_optimizer.py:535: in consolidate_state_dict
    _ = _broadcast_object(
../local/miniconda3/envs/pl-fariscale/lib/python3.9/site-packages/torch/distributed/optim/zero_redundancy_optimizer.py:88: in _broadcast_object
    dist.broadcast(length_tensor, src=src_rank, group=group, async_op=False)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

tensor = tensor([1191]), src = 1
group = <torch._C._distributed_c10d.ProcessGroupNCCL object at 0x7f6e61ec3870>
async_op = False

    def broadcast(tensor, src, group=None, async_op=False):
        """
        Broadcasts the tensor to the whole group.
    
        ``tensor`` must have the same number of elements in all processes
        participating in the collective.
    
        Args:
            tensor (Tensor): Data to be sent if ``src`` is the rank of current
                process, and tensor to be used to save received data otherwise.
            src (int): Source rank.
            group (ProcessGroup, optional): The process group to work on. If None,
                the default process group will be used.
            async_op (bool, optional): Whether this op should be an async op
    
        Returns:
            Async work handle, if async_op is set to True.
            None, if not async_op or if not part of the group
    
        """
        _check_single_tensor(tensor, "tensor")
        if _rank_not_in_group(group):
            return
    
        opts = BroadcastOptions()
        opts.rootRank = src
        opts.rootTensor = 0
    
        if group is None or group is GroupMember.WORLD:
            default_pg = _get_default_group()
>           work = default_pg.broadcast([tensor], opts)
E           RuntimeError: Tensors must be CUDA and dense

../local/miniconda3/envs/pl-fariscale/lib/python3.9/site-packages/torch/distributed/distributed_c10d.py:1159: RuntimeError

```

Is this expected?

Reply:
cc @ananthsub

Reply:
this was fixed in #11886